src/mlpack/methods/hdbscan/hdbscan_impl.hpp

```diff
@@ -0,0 +1,763 @@
+  /**
+ * @file hdbscan_impl.hpp
+ * @author Sudhanshu Ranjan
+ *
+ * Implementation of HDBSCAN.
+ */
+#ifndef __MLPACK_METHODS_HDBSCAN_HDBSCAN_IMPL_HPP
+#define __MLPACK_METHODS_HDBSCAN_HDBSCAN_IMPL_HPP
+
+#include "hdbscan.hpp"
+
+namespace mlpack {
+namespace hdbscan {
+
+/**
+ * Construct the HDBSCAN object with the given parameters.
+ */
+template<typename NeighborSearch,
+         typename MetricType,
+         template<typename TreeMetricType,
+                  typename TreeStatType,
+                  typename TreeMatType> class TreeType >
+HDBSCAN<NeighborSearch,
+        MetricType,
+        TreeType>::
+HDBSCAN(const size_t minPoints,
+        bool allowSingleCluster) :
+        minPoints(minPoints),
+        allowSingleCluster(allowSingleCluster)
+{
+  // Nothing to do.
+}
+
+/**
+ * Performs HDBSCAN clustering on the data,returning the list of cluster 
+ * assignments. The assigned clusters have values from 0 onwards. The noise 
+ * points are labelled as (total nuber of points + 1)
+ *
+ * Dcore -> distance of k-th nearest neighbor
+ *          here k is minPoints
+ * 1. Find the dcore of all the points.
+ * 2. Then append dcore to end of the input data (dataWithDcore)
+ * 3. Compute MST using the Dual Tree Borvuka Method.
+ * 4. Sort the obtained MST on basis of length of edges
+ *    in ascending order.
+ * 5. Covert MST to single likage tree .
+ *    Single linkage tree is a clustering method in which
+ *    we go on clustering the points until a single cluster
+ *    is formed.
+ * 6. Condense the obtained single linkage tree
+ *    to form condensed tree. Condensed tree contains only those
+ *    clusters as parent which have more then minPoints number of
+ *    points.
+ * 7. From this condense tree obtain labels of all the points.
+ */
+template<typename NeighborSearch,
+         typename MetricType,
+         template<typename TreeMetricType,
+                  typename TreeStatType,
+                  typename TreeMatType> class TreeType >
+template<typename MatType>
+void HDBSCAN<NeighborSearch,
+    MetricType,
+    TreeType>::
+Cluster(const MatType& data,
+        arma::Row<size_t>& assignments)
+{
+  // genearting a matrix with the values of dcore
+  arma::Mat<size_t> neighbors;
+  arma::mat distances;
+  neighborSearch.Train(data);
+  neighborSearch.Search(minPoints, neighbors, distances);
+  arma::mat dataWithDcore = arma::conv_to<arma::mat>::from(data);
+  dataWithDcore.resize(dataWithDcore.n_rows+1, dataWithDcore.n_cols);
+  dataWithDcore.row(dataWithDcore.n_rows-1) = distances.row(distances.n_rows-1);
```

Comment:
I see that what you are doing here is a clever trick to make sure that the HDBSCANMetric properly has access to d_core.  But each step of the conversion here causes a runtime penalty, and it seems a little strange to me to force the data the user gives (which could have non-double type) into a double. Instead what I might suggest here is a little bit different... let me know what you think:

Currently the HdbscanMetric class does not hold anything internally.  I suggest adding some instantiated members to the class:

```
MatType& data;
arma::mat coreDistances;
```

Then, for the DualTreeBoruvka class, the data that you call it with is created like this:

```
// Fill a vector with increasing indices.
arma::Row<size_t> mstData = arma::linspace<arma::Row<size_t>>(0,
    data.n_cols - 1, data.n_cols);
```

Then, you could modify HdbscanMetric::Evaluate() to look more like the following:

```
double Evaluate(const VecTypeA& a, const VecTypeB& b)
{
  const size_t aIndex = a[0];
  const size_t bIndex = b[0];

  const double dCoreA = coreDistances(0, aIndex);
  const double dCoreB = coreDistances(0, bIndex);

  const double dist =
      SquaredEuclideanDistance::Evaluate(data.col(aIndex), data.col(bIndex));
  ...
}
```

In this way we can preserve the correct type of the data and avoid some expensive copies.  Let me know what you think of this idea, or if you'd like me to clarify any of my thoughts.